golang/connection.go

```diff
@@ -0,0 +1,475 @@
+		err := msg.read(msgBuf)
+		c.framePool.Release(frame)
+		return err
+	}
+}
+
+// Reserves the next available message id for this connection
+func (c *TChannelConnection) NextMessageId() uint32 {
+	return atomic.AddUint32(&c.nextMessageId, 1)
+}
+
+// Handles a connection error
+func (c *TChannelConnection) connectionError(err error) error {
+	doClose := false
+	c.withStateLock(func() error {
+		if c.state != connectionClosed {
+			c.state = connectionClosed
+			doClose = true
+		}
+		return nil
+	})
+
+	if doClose {
+		c.closeNetwork()
+	}
+
+	return err
+}
+
+// Closes the network connection and all network-related channels
+func (c *TChannelConnection) closeNetwork() {
+	// NB(mmihic): The sender goroutine	will exit once the connection is closed; no need to close
+	// the send channel (and closing the send channel would be dangerous since other goroutine might be sending)
+	if err := c.conn.Close(); err != nil {
+		c.log.Warning("could not close connection to peer %s: %v", c.remotePeerInfo, err)
+	}
+}
+
+// Performs an action with the connection state mutex locked
+func (c *TChannelConnection) withStateLock(f func() error) error {
+	c.stateMut.Lock()
+	defer c.stateMut.Unlock()
+
+	return f()
+}
+
+// Performs an action with the connection state mutex held in a read lock
+func (c *TChannelConnection) withStateRLock(f func() error) error {
+	c.stateMut.RLock()
+	defer c.stateMut.RUnlock()
+
+	return f()
+}
+
+// Main loop that reads frames from the network connection and dispatches to the appropriate handler.
+// Run within its own goroutine to prevent overlapping reads on the socket.  Most handlers simply
+// send the incoming frame to a channel; the init handlers are a notable exception, since we cannot
+// process new frames until the initialization is complete.
+func (c *TChannelConnection) readFrames() {
+	fhBuf := typed.NewReadBufferWithSize(FrameHeaderSize)
+
+	for {
+		if _, err := fhBuf.FillFrom(c.conn, FrameHeaderSize); err != nil {
+			c.connectionError(err)
+			return
+		}
+
+		frame := c.framePool.Get()
+		if err := frame.Header.read(fhBuf); err != nil {
+			// TODO(mmihic): Should be a protocol error
+			c.connectionError(err)
+			return
+		}
+
+		c.log.Info("Recvd: id=%d:type=%d:sz=%d", frame.Header.Id, frame.Header.Type, frame.Header.Size)
+
+		if _, err := c.conn.Read(frame.SizedPayload()); err != nil {
+			c.connectionError(err)
+			return
+		}
+
+		c.log.Info("Rcvd: %s", hex.EncodeToString(frame.SizedPayload()))
+
+		switch frame.Header.Type {
+		case MessageTypeCallReq:
+			c.inbound.handleCallReq(frame)
+		case MessageTypeCallReqContinue:
+			c.inbound.handleCallReqContinue(frame)
+		case MessageTypeCallRes:
+			c.outbound.handleCallRes(frame)
+		case MessageTypeCallResContinue:
+			c.outbound.handleCallResContinue(frame)
+		case MessageTypeInitReq:
+			c.handleInitReq(frame)
+		case MessageTypeInitRes:
+			c.handleInitRes(frame)
+		case MessageTypeError:
+			c.handleError(frame)
```

Comment:
Looking at the TODO list, I don't see anything about protocol-level ping/pong handling. Am I missing where that is implemented?